gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.validator;
+
+import com.google.common.collect.Multimaps;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.ArrayList;
+import org.locationtech.spatial4j.distance.DistanceUtils;
+import org.locationtech.spatial4j.shape.Point;
+import org.locationtech.spatial4j.shape.Shape;
+import org.locationtech.spatial4j.shape.ShapeFactory;
+import org.locationtech.spatial4j.shape.SpatialRelation;
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.annotation.Inject;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsLocationType;
+import org.mobilitydata.gtfsvalidator.table.GtfsShape;
+import org.mobilitydata.gtfsvalidator.table.GtfsShapeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStop;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTime;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTimeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.table.GtfsTripTableContainer;
+import org.mobilitydata.gtfsvalidator.notice.StopTooFarFromTripShapeNotice;
+
+import static org.locationtech.spatial4j.context.SpatialContext.GEO;
+
+/**
+ * Validates that a stop is within a distance threshold for a trip shape.
+ * <p>
+ * Generated notices:
+ * * StopTooFarFromTripShapeNotice
+ */
+@GtfsValidator
+public class StopTooFarFromTripShapeValidator extends FileValidator {
+    @Inject GtfsStopTimeTableContainer stopTimeTable;
+
+    @Inject GtfsTripTableContainer tripTable;
+
+    @Inject GtfsShapeTableContainer shapeTable;
+
+    @Inject GtfsStopTableContainer stopTable;
+
+    // Spatial operation buffer values
+    public static final double TRIP_BUFFER_METERS = 100;
+    private static final double TRIP_BUFFER_DEGREES = DistanceUtils.KM_TO_DEG * (TRIP_BUFFER_METERS / 1000.0d);
+
+    @Override
+    public void validate(NoticeContainer noticeContainer) {
+        // Cache for previously tested shape_id and stop_id pairs - no need to test them more than once.
+        final Map<String, List<String>> testedCache = new HashMap<>();
```

Comment:
Have you considered any alternative data structures? What might the pros/cons be?

Reply:
Yes, I previously used Set<String> to record the shape_id and stop_id pair by connecting them together as a long string key. But problem could happen. For example, when shape_id is "ABC" and stop_id is "DEF", this pair is considered the same as shape_id is "AB" and stop_id is "CDEF", but they are actually different. Because shape_id and stop_id both do not have length limits or requirements, so no matter what we add inside their connecting combination, for example, add "STOP" before stop_id, I don't think it can logically handle this problem. 
Then I also considered to continue using the Set, but maybe transfer shape_id and stop_id into unique hash code and then "connect" the numbers together.
Compared to using Map<String, List<String>> (as what I do here), the advantage of using hash code Set is the storage structure is simpler, but it asks to transfer the ids into hash code first and then compare every time. For the Map, the disadvantage is it is "2D" rather than "1D" as a Set - which means we need to compare key and then the corresponding value each time, but we could compare whether the pair exists and add them at the same time, and there is no need to do further complex calculation. That's why I finally choose to use the Map rather than the Set.

Reply:
Yes. I now understand what you mean. I'm going to change List<String> to Set<String> to have smaller time complexity when checking the uniqueness of stop_id inside.

Reply:
Done!